sunkit_image/granule.py

```diff
@@ -0,0 +1,391 @@
+    fac_pix_limit = fac_size_limit / res
+    fac_brightness_limit = np.mean(data) + 0.5 * np.std(data)
+
+    if len(np.unique(segmented_image)) > 3:
+        raise ValueError('segmented_image must have only values of 1, 0, ' +
+                         'an 0.5 (if dim centers marked)')
+
+    segmented_image = segmented_image
+    segmented_image_fixed = np.copy(segmented_image.astype(float))
+    labeled_seg = skimage.measure.label(segmented_image + 1, connectivity=2)
+    values = np.unique(labeled_seg)
+    for value in values:
+        mask = np.zeros_like(segmented_image)
+        mask[labeled_seg == value] = 1
+        # check that is a 1 (white) region
+        if np.sum(np.multiply(mask, segmented_image)) > 0:
+            region_size = len(segmented_image_fixed[mask == 1])
+            tot_flux = np.sum(data[mask == 1])
+            # check that region is small
+            if region_size < fac_pix_limit:
+                # check that avg flux very high
+                if tot_flux / region_size > fac_brightness_limit:
+                    segmented_image_fixed[mask == 1] = 1.5
+
+    return segmented_image_fixed
+
+def kmeans_segment(data, llambda_axis=-1):
+    """kmeans clustering: uses a kmeans algorithm to cluster data,
+       in order to independently verify the skimage clustering method
+       (e.g using cross_correlation() below).
+        ----------
+       Parameters:
+            data (numpy array): data to be clustered
+            llambda_axis (int): index for wavelength, -1 if scalar array.
+        ----------
+        Returns:
+            labels (numpy array): an array of labels, with 0 = granules,
+                                  2 = intergranules, 1 = in-between.
+            """
+
+    if llambda_axis not in [-1, 2]:
+        raise Exception('Wrong data shape. \
+        (either scalar or (x, y, llambda) )')
+    n_clusters = 3
+    n_init = 20
+    x_size = np.shape(data)[0]
+    y_size = np.shape(data)[1]
+    if llambda_axis == -1:
+        data_flat = np.reshape(data, (x_size * y_size, 1))
+        labels_flat = KMeans(n_clusters).fit(data_flat).labels_
+        labels = np.reshape(labels_flat, (x_size, y_size))
+    else:
+        llambda_size = np.shape(data)[llambda_axis]
+        data = np.reshape(data, (x_size * y_size, llambda_size))
+        labels = np.reshape(Kmeans(n_clusters, n_init).fit(data),
+                            (x_size, y_size))
+
+    # make intergranules 0, granules 1:
+    group0_mean = np.mean(data[labels == 0])
+    group1_mean = np.mean(data[labels == 1])
+    group2_mean = np.mean(data[labels == 2])
+
+    # intergranules
+    min_index = np.argmin([group0_mean,
+                           group1_mean,
+                           group2_mean])
+    segmented_map = np.ones(labels.shape)
+
+    segmented_map[[labels[:, :] == min_index][0]] -= 1
+
+    return segmented_map
+
+def cross_correlation(segment1, segment2):

```

Comment:
maybe call this segment_correlation or segment_overlap?
I think you are just calculating overlap between different segment masks, right?
Cross correlation has a specific meaning in Fourier analysis or other mathematical methods and using that term here could be confusing.

Reply:
Hmm.. I will check with James. I assumed the reason he named it as such was becasue it did indeed compute the actual cross-correlation, but you are right, it really doesn't.